File: iOS/AITranslate/Sources/Services/OfflineModelManager.swift
import Foundation
import Combine
import Translation

/// Status of a model download
enum DownloadStatus: Equatable {
    case idle
    case downloading(progress: Double)
    case processing
    case completed
    case failed(error: String)

    var isDownloading: Bool {
        if case .downloading = self { return true }
        return false
    }

    var isCompleted: Bool {
        self == .completed
    }
}

/// Manages downloading, storing, and accessing offline AI models
@Observable
final class OfflineModelManager {
    static let shared = OfflineModelManager()

    // MARK: - Published State

    private(set) var downloadedPackage: OfflineModelPackage?
    private(set) var downloadProgress: Double = 0.0
    private(set) var downloadStatus: DownloadStatus = .idle
    private(set) var currentlyDownloading: OfflineModelPackage?

    // MARK: - Private Properties

    private let fileManager = FileManager.default

    // MARK: - Initialization

    private init() {
        loadDownloadedState()
    }

    // MARK: - Public Methods

    /// Download an offline model package
    func downloadPackage(_ package: OfflineModelPackage) async throws {
        guard package != .none else { return }

        await MainActor.run {
            currentlyDownloading = package
            downloadStatus = .downloading(progress: 0)
            downloadProgress = 0
        }

        do {
            // Download Whisper model if package includes one
            if let whisperModel = package.whisperModel {
                let whisperKitSize = whisperModel.whisperKitModelSize

                #if DEBUG
                print("[OfflineModelManager] Downloading package: \(package.displayName)")
                print("[OfflineModelManager] Whisper model: \(whisperKitSize.rawValue)")
                #endif

                await MainActor.run {
                    downloadStatus = .downloading(progress: 0.1)
                    downloadProgress = 0.1
                }

                // Download Whisper model via WhisperKit
                try await WhisperKitTranscriber.shared.loadModel(whisperKitSize, download: true)

                await MainActor.run {
                    downloadStatus = .downloading(progress: 0.8)
                    downloadProgress = 0.8
                }
            }

            // If package includes Apple Translation, prompt user to download languages
            // Note: Apple Translation languages are managed by the system
            // Users download them via Settings > General > Language & Region > Translation Languages
            if package.includesAppleTranslation {
                #if DEBUG
                print("[OfflineModelManager] Package includes Apple Translation - languages managed by system")
                #endif
            }

            await MainActor.run {
                downloadStatus = .completed
                downloadProgress = 1.0
                currentlyDownloading = nil
                downloadedPackage = package
            }

            // Save state
            UserDefaults.standard.selectedOfflinePackage = package
            UserDefaults.standard.hasDownloadedModels = true

            #if DEBUG
            print("[OfflineModelManager] Package download completed: \(package.displayName)")
            #endif

        } catch {
            await MainActor.run {
                downloadStatus = .failed(error: error.localizedDescription)
                currentlyDownloading = nil
            }
            throw OfflineModelError.downloadFailed
        }
    }

    /// Delete all downloaded models
    func deleteAllModels() async throws {
        // Unload WhisperKit model
        await WhisperKitTranscriber.shared.unloadModel()

        // Clear state
        await MainActor.run {
            downloadedPackage = nil
            downloadStatus = .idle
            downloadProgress = 0
        }

        UserDefaults.standard.selectedOfflinePackage = nil
        UserDefaults.standard.hasDownloadedModels = false

        #if DEBUG
        print("[OfflineModelManager] All models deleted")
        #endif
    }

    /// Check if any Whisper model is ready for transcription
    func canRecognizeSpeechOffline() -> Bool {
        return WhisperKitTranscriber.shared.modelState.isLoaded
    }

    /// Check if offline translation is available
    /// For Essential/Standard: Only English translation (via Whisper)
    /// For Full Offline/Maximum Quality: Multi-language (via Apple Translation)
    func canTranslateOffline(to targetLanguage: String) -> Bool {
        guard canRecognizeSpeechOffline() else { return false }

        // Whisper can always translate to English
        if targetLanguage == "en" {
            return true
        }

        // For other languages, need Apple Translation (Full Offline or Maximum Quality package)
        guard let package = downloadedPackage, package.includesAppleTranslation else {
            return false
        }

        // Apple Translation availability is checked at runtime
        return true
    }

    /// Get the current package's capabilities description
    var currentCapabilities: String {
        guard let package = downloadedPackage else {
            return "No offline models downloaded"
        }

        switch package {
        case .essential:
            return "Speech recognition (basic) + English translation"
        case .standard:
            return "Speech recognition (good) + English translation"
        case .fullOffline:
            return "Speech recognition (good) + 20 language translation"
        case .maximumQuality:
            return "Speech recognition (best) + 20 language translation"
        case .none:
            return "No offline models downloaded"
        }
    }

    /// Get total estimated size of downloaded models
    func totalDownloadedSizeFormatted() -> String {
        guard let package = downloadedPackage else {
            return "0 MB"
        }

        let formatter = ByteCountFormatter()
        formatter.countStyle = .file
        return formatter.string(fromByteCount: package.totalSizeBytes)
    }

    /// Cancel ongoing download
    func cancelDownload() {
        downloadStatus = .idle
        downloadProgress = 0
        currentlyDownloading = nil
    }

    // MARK: - Private Methods

    private func loadDownloadedState() {
        // Check if we have a saved package
        if let savedPackage = UserDefaults.standard.selectedOfflinePackage {
            // Verify Whisper is still loaded
            if WhisperKitTranscriber.shared.modelState.isLoaded {
                downloadedPackage = savedPackage
            } else {
                // Model was unloaded (app restart), clear state
                downloadedPackage = nil
                UserDefaults.standard.selectedOfflinePackage = nil
            }
        }
    }
}

// MARK: - WhisperModelSize Extension

extension WhisperModelSize {
    /// Convert to WhisperKit model size
    var whisperKitModelSize: WhisperKitTranscriber.ModelSize {
        switch self {
        case .tiny: return .tiny
        case .small: return .small
        case .medium: return .medium
        }
    }
}

// MARK: - Errors

enum OfflineModelError: LocalizedError {
    case downloadFailed
    case modelNotFound

    var errorDescription: String? {
        switch self {
        case .downloadFailed:
            return "Failed to download model. Please check your internet connection and try again."
        case .modelNotFound:
            return "Model not found"
        }
    }
}

// MARK: - UserDefaults Keys

extension UserDefaults {
    private enum Keys {
        static let selectedOfflinePackage = "selectedOfflinePackage"
        static let hasDownloadedModels = "hasDownloadedModels"
    }

    var selectedOfflinePackage: OfflineModelPackage? {
        get {
            guard let rawValue = string(forKey: Keys.selectedOfflinePackage) else { return nil }
            return OfflineModelPackage(rawValue: rawValue)
        }
        set {
            set(newValue?.rawValue, forKey: Keys.selectedOfflinePackage)
        }
    }

    var hasDownloadedModels: Bool {
        get { bool(forKey: Keys.hasDownloadedModels) }
        set { set(newValue, forKey: Keys.hasDownloadedModels) }
    }
}

File: android/app/src/main/java/com/kreativekoala/sayitai/util/OfflineModelManager.kt
package com.kreativekoala.sayitai.util

import android.content.Context
import android.content.SharedPreferences
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.text.DecimalFormat
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Status of a model download
 */
sealed class DownloadStatus {
    object Idle : DownloadStatus()
    data class Downloading(val progress: Double) : DownloadStatus()
    object Processing : DownloadStatus()
    object Completed : DownloadStatus()
    data class Failed(val error: String) : DownloadStatus()

    val isDownloading: Boolean
        get() = this is Downloading

    val isCompleted: Boolean
        get() = this == Completed
}

/**
 * Model type identifier
 */
enum class OfflineModelType(val id: String) {
    WHISPER_TINY("whisper-tiny"),
    WHISPER_SMALL("whisper-small"),
    WHISPER_MEDIUM("whisper-medium"),
    NLLB("nllb-200");

    val displayName: String
        get() = when (this) {
            WHISPER_TINY -> "Whisper Tiny"
            WHISPER_SMALL -> "Whisper Small"
            WHISPER_MEDIUM -> "Whisper Medium"
            NLLB -> "NLLB Translation"
        }

    val fileName: String
        get() = when (this) {
            WHISPER_TINY -> "whisper-tiny.bin"
            WHISPER_SMALL -> "whisper-small.bin"
            WHISPER_MEDIUM -> "whisper-medium.bin"
            NLLB -> "nllb-200-distilled.bin"
        }

    val downloadSizeBytes: Long
        get() = when (this) {
            WHISPER_TINY -> 75_000_000L
            WHISPER_SMALL -> 500_000_000L
            WHISPER_MEDIUM -> 1_500_000_000L
            NLLB -> 600_000_000L
        }

    val downloadSizeFormatted: String
        get() = formatBytes(downloadSizeBytes)

    /**
     * Base URL for model downloads (placeholder - replace with actual CDN)
     */
    val downloadURL: String
        get() = "https://models.sayitai.app/$id/$fileName"

    companion object {
        private fun formatBytes(bytes: Long): String {
            if (bytes < 1024) return "$bytes B"
            val kb = bytes / 1024.0
            if (kb < 1024) return "${DecimalFormat("#.#").format(kb)} KB"
            val mb = kb / 1024.0
            if (mb < 1024) return "${DecimalFormat("#.#").format(mb)} MB"
            val gb = mb / 1024.0
            return "${DecimalFormat("#.#").format(gb)} GB"
        }
    }
}

/**
 * Manages downloading, storing, and accessing offline AI models
 */
@Singleton
class OfflineModelManager @Inject constructor(
    @ApplicationContext private val context: Context
) {
    private val prefs: SharedPreferences = context.getSharedPreferences("offline_models", Context.MODE_PRIVATE)

    // StateFlow for observable state
    private val _downloadedWhisperModel = MutableStateFlow<WhisperModelSize?>(null)
    val downloadedWhisperModel: StateFlow<WhisperModelSize?> = _downloadedWhisperModel.asStateFlow()

    private val _isNLLBDownloaded = MutableStateFlow(false)
    val isNLLBDownloaded: StateFlow<Boolean> = _isNLLBDownloaded.asStateFlow()

    private val _downloadProgress = MutableStateFlow(0.0)
    val downloadProgress: StateFlow<Double> = _downloadProgress.asStateFlow()

    private val _downloadStatus = MutableStateFlow<DownloadStatus>(DownloadStatus.Idle)
    val downloadStatus: StateFlow<DownloadStatus> = _downloadStatus.asStateFlow()

    private val _currentlyDownloading = MutableStateFlow<OfflineModelType?>(null)
    val currentlyDownloading: StateFlow<OfflineModelType?> = _currentlyDownloading.asStateFlow()

    private val modelsDirectory: File
        get() {
            val dir = File(context.filesDir, "OfflineModels")
            if (!dir.exists()) {
                dir.mkdirs()
            }
            return dir
        }

    init {
        loadDownloadedModels()
    }

    // MARK: - Public Methods

    /**
     * Download an offline model package
     */
    suspend fun downloadPackage(modelPackage: OfflineModelPackage) {
        if (modelPackage == OfflineModelPackage.NONE) return

        val modelsToDownload = mutableListOf<OfflineModelType>()

        modelPackage.whisperModel?.let { whisperModel ->
            when (whisperModel) {
                WhisperModelSize.TINY -> modelsToDownload.add(OfflineModelType.WHISPER_TINY)
                WhisperModelSize.SMALL -> modelsToDownload.add(OfflineModelType.WHISPER_SMALL)
                WhisperModelSize.MEDIUM -> modelsToDownload.add(OfflineModelType.WHISPER_MEDIUM)
            }
        }

        if (modelPackage.includesNLLB) {
            modelsToDownload.add(OfflineModelType.NLLB)
        }

        for (model in modelsToDownload) {
            downloadModel(model)
        }

        loadDownloadedModels()
    }

    /**
     * Download a specific model
     */
    suspend fun downloadModel(modelType: OfflineModelType) {
        _currentlyDownloading.value = modelType
        _downloadStatus.value = DownloadStatus.Downloading(0.0)
        _downloadProgress.value = 0.0

        try {
            withContext(Dispatchers.IO) {
                val url = URL(modelType.downloadURL)
                val connection = url.openConnection() as HttpURLConnection
                connection.requestMethod = "GET"
                connection.connect()

                if (connection.responseCode != HttpURLConnection.HTTP_OK) {
                    throw OfflineModelException.DownloadFailed("HTTP ${connection.responseCode}")
                }

                val fileLength = connection.contentLengthLong
                val destinationFile = File(modelsDirectory, modelType.fileName)

                // Delete existing file if present
                if (destinationFile.exists()) {
                    destinationFile.delete()
                }

                connection.inputStream.use { input ->
                    FileOutputStream(destinationFile).use { output ->
                        val buffer = ByteArray(8192)
                        var totalBytesRead = 0L
                        var bytesRead: Int

                        while (input.read(buffer).also { bytesRead = it } != -1) {
                            output.write(buffer, 0, bytesRead)
                            totalBytesRead += bytesRead

                            if (fileLength > 0) {
                                val progress = totalBytesRead.toDouble() / fileLength.toDouble()
                                _downloadProgress.value = progress
                                _downloadStatus.value = DownloadStatus.Downloading(progress)
                            }
                        }
                    }
                }
            }

            _downloadStatus.value = DownloadStatus.Processing

            // Brief processing delay
            withContext(Dispatchers.IO) {
                Thread.sleep(200)
            }

            _downloadStatus.value = DownloadStatus.Completed
            _currentlyDownloading.value = null
            loadDownloadedModels()

        } catch (e: Exception) {
            _downloadStatus.value = DownloadStatus.Failed(e.message ?: "Unknown error")
            _currentlyDownloading.value = null
            throw e
        }
    }

    /**
     * Delete a downloaded model
     */
    fun deleteModel(modelType: OfflineModelType) {
        val modelFile = File(modelsDirectory, modelType.fileName)
        if (modelFile.exists()) {
            modelFile.delete()
        }
        loadDownloadedModels()
    }

    /**
     * Delete all downloaded models
     */
    fun deleteAllModels() {
        modelsDirectory.listFiles()?.forEach { file ->
            file.delete()
        }
        loadDownloadedModels()
    }

    /**
     * Check if a specific model is downloaded
     */
    fun isModelDownloaded(modelType: OfflineModelType): Boolean {
        val modelFile = File(modelsDirectory, modelType.fileName)
        return modelFile.exists()
    }

    /**
     * Get the path to a downloaded model
     */
    fun modelPath(modelType: OfflineModelType): File? {
        val file = File(modelsDirectory, modelType.fileName)
        return if (file.exists()) file else null
    }

    /**
     * Check if offline translation is available for a language pair
     */
    fun canTranslateOffline(sourceCode: String, targetCode: String): Boolean {
        // If we have Whisper, we can transcribe any language
        // If we have NLLB, we can translate between supported languages
        // Without NLLB, Whisper only translates TO English

        if (_downloadedWhisperModel.value == null) return false

        return if (_isNLLBDownloaded.value) {
            // NLLB supports translation between most language pairs
            true
        } else {
            // Whisper alone only translates TO English
            targetCode == "en"
        }
    }

    /**
     * Check if offline speech recognition is available
     */
    fun canRecognizeSpeechOffline(): Boolean {
        return _downloadedWhisperModel.value != null
    }

    /**
     * Get total size of downloaded models
     */
    fun totalDownloadedSize(): Long {
        var totalSize = 0L
        OfflineModelType.entries.forEach { modelType ->
            if (isModelDownloaded(modelType)) {
                totalSize += modelType.downloadSizeBytes
            }
        }
        return totalSize
    }

    /**
     * Get formatted total size string
     */
    fun totalDownloadedSizeFormatted(): String {
        return formatBytes(totalDownloadedSize())
    }

    /**
     * Cancel ongoing download
     */
    fun cancelDownload() {
        _downloadStatus.value = DownloadStatus.Idle
        _downloadProgress.value = 0.0
        _currentlyDownloading.value = null
    }

    // Selected package preference
    var selectedOfflinePackage: OfflineModelPackage?
        get() {
            val id = prefs.getString("selectedOfflinePackage", null) ?: return null
            return OfflineModelPackage.entries.find { it.id == id }
        }
        set(value) {
            prefs.edit().putString("selectedOfflinePackage", value?.id).apply()
        }

    var hasDownloadedModels: Boolean
        get() = prefs.getBoolean("hasDownloadedModels", false)
        set(value) = prefs.edit().putBoolean("hasDownloadedModels", value).apply()

    // MARK: - Private Methods

    private fun loadDownloadedModels() {
        // Check which Whisper model is downloaded (prefer larger)
        _downloadedWhisperModel.value = when {
            isModelDownloaded(OfflineModelType.WHISPER_MEDIUM) -> WhisperModelSize.MEDIUM
            isModelDownloaded(OfflineModelType.WHISPER_SMALL) -> WhisperModelSize.SMALL
            isModelDownloaded(OfflineModelType.WHISPER_TINY) -> WhisperModelSize.TINY
            else -> null
        }

        // Check NLLB
        _isNLLBDownloaded.value = isModelDownloaded(OfflineModelType.NLLB)

        // Update hasDownloadedModels flag
        hasDownloadedModels = _downloadedWhisperModel.value != null || _isNLLBDownloaded.value
    }

    companion object {
        private fun formatBytes(bytes: Long): String {
            if (bytes < 1024) return "$bytes B"
            val kb = bytes / 1024.0
            if (kb < 1024) return "${DecimalFormat("#.#").format(kb)} KB"
            val mb = kb / 1024.0
            if (mb < 1024) return "${DecimalFormat("#.#").format(mb)} MB"
            val gb = mb / 1024.0
            return "${DecimalFormat("#.#").format(gb)} GB"
        }
    }
}

/**
 * Offline model errors
 */
sealed class OfflineModelException(message: String) : Exception(message) {
    object InvalidURL : OfflineModelException("Invalid model download URL")
    class DownloadFailed(details: String) : OfflineModelException("Failed to download model: $details")
    object UnzipFailed : OfflineModelException("Failed to extract model")
    object ModelNotFound : OfflineModelException("Model file not found")
}
